arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EagerEffectScope.kt

```diff
@@ -186,6 +188,38 @@ public interface EagerEffectScope<in R> {
    */
   public suspend fun ensure(condition: Boolean, shift: () -> R): Unit =
     if (condition) Unit else shift(shift())
+
+  /**
+   * Initializes an "attempt block", which encloses an action for which
+   * you want to catch any possible `shift`ed outcome. This should be used
+   * in combination with `catch`.
+   *
+   * ```
+   * attempt { ... } catch { ... }
+   * ```
+   *
+   * The [effect] may `shift` into a different `ErrorScope`, giving
+   * the change to a later `recover` to change the shifted value.
+   * This is useful to simulate re-throwing of exceptions.
+   */
+  @OptIn(ExperimentalTypeInference::class)
+  public suspend fun <E, A> attempt(
+    @BuilderInference
+    f: suspend EagerEffectScope<E>.() -> A,
+  ): suspend EagerEffectScope<E>.() -> A = f
+
+  /**
+   * Finishes an "attempt block" by providing the way in which to [recover]
+   * from a `shift`ed outcome. This function should be used in combination
+   * with `attempt`.
```

Comment:
Maybe this Doc suggestion  is a bit more clear - I kept the attempt terminology. 
```suggestion
   * Runs the receiver within an [eagerEffect] and running [fold] function over the computation.
   *
   * When the [EagerEffect] has shifted with [R] it will [recover] the shifted value to [A], and when it
   * ran the computation to completion it will return the value [A].
   * [catch] is used in combination with [attempt].
```

I would probably prefer a real example below like: 

```
* ```kotlin
 * import arrow.core.Either
 * import arrow.core.None
 * import arrow.core.Option
 * import arrow.core.Validated
 * import arrow.core.continuations.eagerEffect
 * import io.kotest.assertions.fail
 * import io.kotest.matchers.shouldBe
 *
 * fun main() {
 *   eagerEffect<String, Int> {
 *     val x = Either.Right(1).bind()
 *     val y = Validated.Valid(2).bind()
 *     val z = 
 *      attempt { Option(3).bind { "Option was empty" } } catch { 0 }
 *     x + y + z
 *   }.fold({ fail("Shift can never be the result") }, { it shouldBe 6 })
* ```
```

Reply:
Shouldn't this snippet show the behaviour of `catch` in this case?

```kotlin
attempt { None.bind { "Option was empty" } } catch { 0 }
```

Reply:
I changed your comments by your suggestions, but avoided the reference to `fold`, since I feel this is not a part of the API the user needs to understand in order to know what `catch` is doing.